test/PandoraInterface.cxx

```diff
@@ -257,40 +271,58 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+            // Merge voxels with the same IDs
+            std::vector<LArVoxel> mergedVoxels = MergeSameVoxels(voxelList);
+
+            std::cout << "Produced " << mergedVoxels.size() << " merged voxels from " << voxelList.size() << " voxels." << std::endl;
+
+            // Loop over the voxels and make them into caloHits
+            for (int i = 0; i < mergedVoxels.size(); i++)
             {
-                // const double length = g4Hit->GetTrackLength();
-                // const int contrib = g4Hit->Contrib.front();
-                // const std::string particle =
-                // pEDepSimEvent->Trajectories[contrib].GetName();
-
-                const double energy = g4Hit->GetEnergyDeposit();
-                const TLorentzVector &hitStart = g4Hit->GetStart() * mm2cm;
-                const TLorentzVector &hitStop = g4Hit->GetStop() * mm2cm;
-                const CartesianVector start(hitStart.X(), hitStart.Y(), hitStart.Z());
-                const CartesianVector stop(hitStop.X(), hitStop.Y(), hitStop.Z());
-                const CartesianVector centre((start + stop) * 0.5f);
+                const LArVoxel voxel = mergedVoxels[i];
+                const CartesianVector voxelPos(voxel.m_voxelPosVect);
+                const double voxelE = voxel.m_energyInVoxel;
 
                 lar_content::LArCaloHitParameters caloHitParameters;
-                caloHitParameters.m_positionVector = centre;
+                caloHitParameters.m_positionVector = voxelPos;
                 caloHitParameters.m_expectedDirection = pandora::CartesianVector(0.f, 0.f, 1.f);
                 caloHitParameters.m_cellNormalVector = pandora::CartesianVector(0.f, 0.f, 1.f);
                 caloHitParameters.m_cellGeometry = pandora::RECTANGULAR;
-                caloHitParameters.m_cellSize0 = std::fabs(start.GetY() - stop.GetY());
-                caloHitParameters.m_cellSize1 = std::fabs(start.GetX() - stop.GetX());
-                caloHitParameters.m_cellThickness = std::fabs(start.GetZ() - stop.GetZ());
+                caloHitParameters.m_cellSize0 = voxelWidth;
+                caloHitParameters.m_cellSize1 = voxelWidth;
+                caloHitParameters.m_cellThickness = voxelWidth;
                 caloHitParameters.m_nCellRadiationLengths = 1.f;
                 caloHitParameters.m_nCellInteractionLengths = 1.f;
                 caloHitParameters.m_time = 0.f;
-                caloHitParameters.m_inputEnergy = energy;
-                caloHitParameters.m_mipEquivalentEnergy = energy;
-                caloHitParameters.m_electromagneticEnergy = energy;
-                caloHitParameters.m_hadronicEnergy = energy;
+                caloHitParameters.m_inputEnergy = voxelE;
+                caloHitParameters.m_mipEquivalentEnergy = voxelE;
+                caloHitParameters.m_electromagneticEnergy = voxelE;
+                caloHitParameters.m_hadronicEnergy = voxelE;
```

Comment:
What are the units of energy here?

Reply:
Energy is GeV, which is converted from the G4 hit energy in MeV within `MakeVoxels()`.